src/otaclient/grpc/api_v2/servicer.py

```diff
@@ -64,20 +68,57 @@ def __init__(
 
         self._op_queue = op_queue
         self._resp_queue = resp_queue
+        self._main_queue = main_queue
 
         self._ecu_status_storage = ecu_status_storage
+        self._critical_zone_flags = critical_zone_flags
         self._polling_waiter = self._ecu_status_storage.get_polling_waiter()
 
     def _local_update(self, request: UpdateRequestV2) -> api_types.UpdateResponseEcu:
         """Thread worker for dispatching a local update."""
         return self._dispatch_local_request(request, api_types.UpdateResponseEcu)
 
+    def _local_stop(self, request: StopRequestV2) -> api_types.StopResponseEcu:
+        """Thread worker for dispatching a local stop request."""
+        return self._dispatch_local_stop_request(request, api_types.StopResponseEcu)
+
     def _local_client_update(
         self, request: ClientUpdateRequestV2
     ) -> api_types.ClientUpdateResponseEcu:
         """Thread worker for dispatching a local client update."""
         return self._dispatch_local_request(request, api_types.ClientUpdateResponseEcu)
 
+    def _dispatch_local_stop_request(
+        self,
+        request: StopRequestV2,
+        response_type: type[api_types.StopResponseEcu],
+    ) -> api_types.StopResponseEcu:
+        """Dispatch stop request to main process."""
+        try:
+            # Check if critical zone is available
+            if not self._critical_zone_flags.is_critical_zone.acquire(block=False):
+                logger.error("Going through critical zone, rejecting stop request")
+                return response_type(
+                    ecu_id=self.my_ecu_id,
+                    result=api_types.FailureType.RECOVERABLE,
+                    message="In critical zone, stop request rejected",
+                )
```

Comment:
Could also use context manager following my suggestion of the `CriticalZoneFlag` implementation.